### PR TITLE
[ROCKETMQ-335] Reload server certificate, private key and root ca when these are changed

### PR DESCRIPTION
**Pull request by zhouxinyu**
## What is the purpose of the change

JIRA: https://issues.apache.org/jira/browse/ROCKETMQ-335

TLS is supported in 4.2.0, but it doesn't support reload server certificate, private key and root ca when these are changed.

This feature can allow us to update the TLS pem files without downtime.

## Brief changelog

1. Add a FileWatchService to watch a target file when it's changed.
2. Support reload SSLContext for Broker and NameServer.

The multiple commits will be squashed when this PR is merged.

## Verifying this change

1. Run the FileWatchServiceTest to verify the notify mechanism for changed file.
2. Run the unit test `reloadSslContextForServer` to verify the SSLContext reload mechanism.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/ROCKETMQ/issues/) filed for the change (usually before you start working on it). Trivial changes like typos do not require a JIRA issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ROCKETMQ-XXX] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [x] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: broker/src/main/java/org/apache/rocketmq/broker/BrokerController.java [hunk 1]
```diff
@@ -387,6 +391,45 @@ public void run() {
...
+                                    log.info("The trust certificate changed, reload the ssl context");
+                                    reloadServerSslContext();
+                                }
+                                if (path.equals(TlsSystemConfig.tlsServerCertPath)) {
+                                    certChanged = true;
+                                }
+                                if (path.equals(TlsSystemConfig.tlsServerKeyPath)) {
+                                    keyChanged = true;
+                                }
+                                if (certChanged && keyChanged) {
```
Comment (lizhanhui) [inline broker/src/main/java/org/apache/rocketmq/broker/BrokerController.java]: It's possible to just change the certificate file

File: namesrv/src/main/java/org/apache/rocketmq/namesrv/NamesrvController.java [hunk 2]
```diff
@@ -95,6 +99,44 @@ public void run() {
...
+                                reloadServerSslContext();
+                            }
+                            if (path.equals(TlsSystemConfig.tlsServerCertPath)) {
+                                certChanged = true;
+                            }
+                            if (path.equals(TlsSystemConfig.tlsServerKeyPath)) {
+                                keyChanged = true;
+                            }
+                            if (certChanged && keyChanged) {
+                                log.info("The certificate and private key changed, reload the ssl context");
```
Comment (lizhanhui) [inline namesrv/src/main/java/org/apache/rocketmq/namesrv/NamesrvController.java]: Same here.

File: broker/src/main/java/org/apache/rocketmq/broker/BrokerController.java (hunk 1 above)
Comment (zhouxinyu) [reply broker/src/main/java/org/apache/rocketmq/broker/BrokerController.java]: But if both certificate and private key are changed, there may have an inconsistent period which will result in server unavailable.

And in most cases, certificate and private key will be refreshed at the same time. 

BTW, there is a better solution that reloads only when certificate and private key are matched,  but it's more complicated, considering the various security algorithm.

So, this is a tradeoff way, simple and effective.

**Comment by lizhanhui** [comment]
I am OK. +1 to merge it.

**Comment by lollipopjin** [comment]
LGTM，+1.

**Comment by vintagewang** [comment]
+1

**Review by vongosling — APPROVED** [review]
+1

**Pull request merged**